### PR TITLE
Avatar's own view of the world

### PR DESCRIPTION
**Pull request by ramonfmir**
Each avatar now holds an `AvatarView` instance. It calculates the cells to be created and the cells to be deleted after every move action according to the radius specified and the avatar location. This takes full advantage of the updates-based API (PR #219 - merged) by ensuring that there are no redundant objects in the scene, which will become important when we use the [Unity project](https://github.com/ocadotechnology/aimmo-unity/) for the front-end since having a lightweight scene will improve both the loading time and the FPS. Also, this avatar view is a must if we want to have infinitely large maps.

### CODE REVIEW

File: aimmo-game/simulation/avatar/avatar_wrapper.py [hunk 1]
```diff
@@ -2,7 +2,8 @@
 import requests
 
 from simulation.action import ACTIONS, MoveAction, WaitAction
-
+from simulation.location import Location
+from avatar_view import AvatarView
```
Comment (danalex97) [inline aimmo-game/simulation/avatar/avatar_wrapper.py]: Avatar view should be imported absolutely.

File: aimmo-game/simulation/location.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def __ne__(self, other):
         return not self == other
 
     def __hash__(self):
-        return hash((self.x, self.y))
+        return hash(str(self.x) + ":" + str(self.y))
```
Comment (danalex97) [inline aimmo-game/simulation/location.py]: I think the old hashing function would be better.

File: aimmo-game/tests/test_simulation/avatar/test_avatar_view.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
+from unittest import TestCase
+
+from simulation.world_map import Cell
+from simulation.location import Location
+from simulation.direction import EAST, WEST, NORTH, SOUTH
+from simulation.avatar.avatar_view import AvatarView
+
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/avatar/test_avatar_view.py]: two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

File: aimmo-game/tests/test_simulation/avatar/test_avatar_view.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    def max_y(self):
+        return self._max_y
+
+    def get_cell(self, location):
+        if not location is None:
+            return Cell(location)
+        else:
+            raise ValueError
+
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/avatar/test_avatar_view.py]: two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

File: aimmo-game/simulation/avatar/avatar_view.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        # Update cells in view. (Note that these are set operations: union and set difference)
+        self.cells_in_view |= self.cells_to_reveal
+        self.cells_in_view -= self.cells_to_clear
+
+        # Shift the four view corners (horizons).
+        self.NE_horizon += move_direction
+        self.NW_horizon += move_direction
+        self.SE_horizon += move_direction
+        self.SW_horizon += move_direction
```
Comment (CelineBoudier) [inline aimmo-game/simulation/avatar/avatar_view.py]: new line in the end please

File: aimmo-game/simulation/world_state.py [hunk 6]
```diff
@@ -82,40 +83,59 @@ def player_dict(avatar):
...
+            for cell in avatar_view.cells_in_view:
+                if not cell.add_to_scene is None:
+                    self.create_map_feature(cell.add_to_scene.value, map_feature_dict(cell))
+                if not cell.remove_from_scene is None:
+                    self.delete_map_feature(cell.remove_from_scene.value, map_feature_dict(cell))
+
+            # Deletion
+            for cell in avatar_view.cells_to_clear:
+                # There is an avatar.
+                if not cell.avatar is None:
```
Comment (CelineBoudier) [inline aimmo-game/simulation/world_state.py]: or, if cell.avatar is not None: :)

File: aimmo-game/tests/test_simulation/avatar/test_avatar_view.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    def test_chain_of_moves(self):
+        mock_world_map = MockWorldMap(-26, -32, 22, 27)
+        avatar_view = AvatarView(Location(1, 1), 3)
+        avatar_view.reveal_all_cells(mock_world_map)
+        initial_cells_in_view = set(avatar_view.cells_in_view)
+        avatar_view.move(EAST, mock_world_map)
+        avatar_view.move(SOUTH, mock_world_map)
+        avatar_view.move(WEST, mock_world_map)
+        avatar_view.move(NORTH, mock_world_map)
+        self.assertEqual(avatar_view.cells_in_view, initial_cells_in_view)
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/avatar/test_avatar_view.py]: new end line

**Pull request merged**